plugin/evm/customheader/min_delay_excess.go

```diff
@@ -0,0 +1,127 @@
+// See the file LICENSE for licensing terms.
+
+package customheader
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ava-labs/avalanchego/vms/evm/acp226"
+	"github.com/ava-labs/libevm/core/types"
+
+	"github.com/ava-labs/coreth/params/extras"
+	"github.com/ava-labs/coreth/plugin/evm/customtypes"
+)
+
+var (
+	errRemoteMinDelayExcessSet = errors.New("remote min delay excess should be nil")
+	errRemoteMinDelayExcessNil = errors.New("remote min delay excess should not be nil")
+	errIncorrectMinDelayExcess = errors.New("incorrect min delay excess")
+	errParentMinDelayExcessNil = errors.New("parent min delay excess should not be nil")
+)
+
+// MinDelay
+//
+// If the `desiredMinDelayExcess` is nil, the parent's delay excess is used.
+func MinDelayExcess(
+	config *extras.ChainConfig,
+	parent *types.Header,
+	header *types.Header,
+	desiredMinDelayExcess *uint64,
+) (*uint64, error) {
+	if config.IsGranite(header.Time) {
+		minDelayExcess, err := minDelayExcessAfterBlock(config, parent, desiredMinDelayExcess)
+		if err != nil {
+			return nil, fmt.Errorf("calculating min delay excess: %w", err)
+		}
+		minDelayExcessU64 := uint64(minDelayExcess)
+		return &minDelayExcessU64, nil
+	} else { // Prior to Granite there was no expected min delay excess.
+		return nil, nil
+	}
+}
+
+// VerifyMinDelayExcess verifies that the min delay excess in header is consistent.
+func VerifyMinDelayExcess(
+	config *extras.ChainConfig,
+	parent *types.Header,
+	header *types.Header,
+) error {
+	switch {
+	case config.IsGranite(header.Time):
+		remoteDelayExcess := customtypes.GetHeaderExtra(header).MinDelayExcess
+		if remoteDelayExcess == nil {
+			return fmt.Errorf("%w: %s", errRemoteMinDelayExcessNil, header.Hash())
+		}
+		// By passing in the claimed excess, we ensure that the expected
+		// excess is equal to the claimed excess if it is possible
+		// to have correctly set it to that value. Otherwise, the resulting
+		// value will be as close to the claimed value as possible, but would
+		// not be equal.
+		expectedDelayExcess, err := minDelayExcessAfterBlock(
+			config,
+			parent,
+			remoteDelayExcess,
+		)
+		if err != nil {
+			return fmt.Errorf("calculating expected min delay excess: %w", err)
+		}
+
+		if *remoteDelayExcess != uint64(expectedDelayExcess) {
+			return fmt.Errorf("%w: expected %+v, found %+v",
+				errIncorrectMinDelayExcess,
+				expectedDelayExcess,
+				remoteDelayExcess,
+			)
+		}
+	default:
+		// Prior to Granite there was no expected min delay excess.
+		// TODO (ceyonur): this can be removed after Granite is activated.
+		if customtypes.GetHeaderExtra(header).MinDelayExcess != nil {
+			return fmt.Errorf("%w: %s", errRemoteMinDelayExcessSet, header.Hash())
+		}
+	}
+	return nil
+}
+
+// minDelayExcessBeforeBlock takes the previous header and the timestamp of its child
+// block and calculates the minimum delay before the child block is executed.
+// Should only be called with timestamp in Granite.
```

Comment:
This comment seems inaccurate for ACP-226 because it doesn't take the child timestamp.

Reply:
IIUC, this function is really just returning the effective min delay excess of the given header, right?

Reply:
I think we probably don't need this function as in acp-176, so just merged two functions.